### PR TITLE
Merge test execution time across GitHub Actions matrix-sharded jobs

### PR DESCRIPTION
**Pull request by wreulicke**
## Summary
- Add `testExecutionTime.jobs` (glob patterns, e.g. `test (*)`) so octocov can look up the test-running step across other jobs of the same workflow run, not just the job octocov itself runs in.
- This enables `testExecutionTime` to be measured correctly when tests are split across a GitHub Actions matrix (test sharding) and coverage/execution time are aggregated in a separate job, mirroring the existing coverage-merging story documented in the README.
- Without `jobs:` configured, behavior is unchanged: only the job octocov itself runs in is searched (via the existing current-job detection heuristic).
- Job names are matched with the same glob library (`doublestar`) already used elsewhere in octocov (`coverage.exclude`, `codeToTestRatio`), so no new dependency was introduced.

## Test plan
- [x] `go build ./...`
- [x] `go vet ./...`
- [x] `go test ./...`
- [x] `golangci-lint run ./...`
- [x] Added unit tests in `gh/gh_test.go` covering `ResolveJobIDs` (glob matching, no-match error, fallback to current job) and `FetchStepByTimeAcrossJobs` (job filtering by ID)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by wreulicke** [comment]
@k1LoW 
Could you check this PR?

**Comment by k1LoW** [comment]
Sorry for the wait 🙏 
We're currently reviewing compatibility with the existing octocov design.

File: gh/gh.go [hunk 1]
```diff
@@ -453,7 +455,113 @@ type Step struct {
...
+	}
+	p := backoff.Exponential( //nostyle:funcfmt
+		backoff.WithMinInterval(time.Second),
+		backoff.WithMaxInterval(30*time.Second),
+		backoff.WithJitterFactor(0.05),
+		backoff.WithMaxRetries(5),
+	)
+	b := p.Start(ctx)
+	for backoff.Continue(b) {
+		jobs, _, err := g.client.Actions.ListWorkflowJobs(ctx, owner, repo, runID, &github.ListWorkflowJobsOptions{})
```
Comment (wreulicke) [inline gh/gh.go]: Current implementation calls ListWorkflowJobs API 3 times in ResolveJobIDs/FetchStepByTimeAcrossJobs/FetchStepsByName.
It is not efficient...

**Comment by wreulicke** [comment]
I realized that this doesn't quite meet my requirements.

I want to configure jobs without specifying stepNames. However, with the current design, if stepNames is omitted, it finds the steps based on the creation time of the coverage report files. Because of this, we would still need to update .octocov.yml even if this is merged.

That said, changing the current design would break backward compatibility. Therefore, I will update the settings after this change is released.

**Comment by k1LoW** [comment]
@wreulicke 
Thanks for the follow-up and for digging into this.

Let me clarify the design intent behind `testExecutionTime`, since I think it explains why the `jobs:`-only approach is hard to make work.

`testExecutionTime` measures **the elapsed time of the step that ran the tests**. The step is the unit of measurement, so the whole feature hinges on identifying *which* step that was. When `steps:` is omitted, octocov identifies it by taking the mtime of the coverage report files it loaded and looking for the step whose execution window contains that timestamp. That works because the coverage report is written by the test step itself, within the same job.

Once you cross the job boundary, that identification breaks down. Even with the mtimes of every collected covPath, `jobs:` + mtime isn't enough to pin down the test step, because matrix jobs run concurrently and their step windows overlap.

For example:

| job | test step window | `lcov.info` mtime |
| --- | --- | --- |
| `test (1)` | `[T, T+30m]` | ~T+30m |
| `test (2)` | `[T, T+10m]` | ~T+10m |

The mtime from `test (2)` (~T+10m) also falls inside `test (1)`'s window, so the timestamp alone cannot tell us which job produced that file. The current implementation returns whichever job the API happens to list first, so `test (1)` wins and `test (2)`'s step is never measured. When shards start at roughly the same time, a shorter shard's window is always contained in a longer one's, so this is the normal case for sharded tests rather than an edge case.

What's missing is a link from a coverage file back to the job that produced it. The GitHub API doesn't expose a `job_id` on artifacts, so there's nothing on the aggregation side to reconstruct it from.

**Comment by wreulicke** [comment]
> Let me clarify the design intent behind `testExecutionTime`, since I think it explains why the `jobs:`-only approach is hard to make work.
> 
> `testExecutionTime` measures **the elapsed time of the step that ran the tests**. The step is the unit of measurement, so the whole feature hinges on identifying _which_ step that was. When `steps:` is omitted, octocov identifies it by taking the mtime of the coverage report files it loaded and looking for the step whose execution window contains that timestamp. That works because the coverage report is written by the test step itself, within the same job.

@k1LoW 
Thank you.
I agree with you completely. 
So, I will consider adding `steps:` to the config after this PR is merged and released.

BTW, it looks upload-artifact does not keep the timestamp by default 😇 ... 
ref: https://github.com/actions/upload-artifact/issues/384

**Comment by wreulicke** [comment]
Hi, @k1LoW.

Could you review this PR again?
I wanna merge test execution time even if config changes is needed.

**Comment by k1LoW** [comment]
@wreulicke

Thank you for requesting a re-review. There are two things I would like to address before we can consider merging this.

### 1. First, could you try setting only testExecutionTime.steps: in the current main (or the latest release) and test it out?

When `steps:` is specified, even with the current implementation, octocov collects steps with the same name across all jobs within a workflow run and sums them up, excluding overlapping time intervals. It does not filter by job. This behavior was introduced in PR #28, "Support measure parallel/multi tests execution time," specifically to sum up the execution time of tests that are run in parallel or split up.

In other words, summing up the execution times of sharded Run test jobs using a matrix should be achievable with just the following configuration, without any code changes.

testExecutionTime:
  steps:
    - Run test

There is no need to wait for this PR to be merged or for a subsequent release. If this does not yield the expected values, please share your workflow configuration along with the actual error message (or the values obtained). That will point us to the exact part that needs to be fixed.

### 2. Supporting only `jobs:` is not theoretically feasible, so the implementation in this PR cannot be accepted

As explained in my previous comment, `jobs:` + mtime cannot identify the step that executed the tests. Because the execution intervals of steps in a matrix job overlap with each other, mtime alone makes it impossible to determine which job generated that file, and the implementation simply picks the first job that matches in the order returned by the API. As a result, the steps of shorter shards are left unmeasured.

Even in this PR, FindStepByTime iterates over the target jobs in the API enumeration order and returns the first matching step, so this ambiguity has not been resolved. Only job filtering has been added.

In addition, in the aggregation flow in the README, the coverage report consists of files extracted by actions/download-artifact on the aggregation job side. Since they are extracted after `needs:` test, their mtime becomes the timestamp after all test jobs have completed. Combined with the issue you pointed out regarding upload-artifact not preserving timestamps, the mtime does not fall within the execution interval of any test step.

Therefore, the implementation in this PR only functions when used in combination with `steps:`, which is the very configuration this PR tries to avoid. And if you are going to use `steps:`, as mentioned in point 1, the current code already achieves your goal.

**Comment by wreulicke** [comment]
@k1LoW 
Oh, I misunderstood...
I will try it lator. Sorry for bothering you.
For now, I will close this PR.

**Comment by k1LoW** [comment]
@wreulicke

No need to apologize at all. `steps:` collecting the step across every job in the run isn't documented anywhere, so this was on the docs, not on you. Thanks for pushing on this and for the detailed write-up along the way.

I'll add a note about the cross-job behavior to the `testExecutionTime.steps` section of the README so this is discoverable.

Thank you!

**Comment by wreulicke** [comment]
@k1LoW 
Adding `steps:`, I tried again.
It works. We can collect test execution time correctly!

image

Thank you!
